// mobstor/test/senarios.js
/*
 * Copyright (c) 2011 Yahoo! Inc. All rights reserved.
 */
//build MObStor mock object
exports.senario = {
    'safeplay.yahoofs.com' : {
        httpVersion : '1.0',
        defaultEncoding : 'utf-8',
        storage : '/tmp',
        certificate : 'yahoo.opsdb.site.corp-snv: v=1;a=yahoo.opsdb.site.corp-snv;h=207.126.232.237;t=1314509085;s=MEo1.gg7nIapd_oB3uT0xeTHxHw0Pj3dNk61.dC0VI_.1_sPhxezHkGXGyoRkkWsUrfp.K_Qo6_7hPUaR51bnQ--',
        senario : [
            {
                request : {
                    method : "PUT",
                    path : "/stream.bin",
                    qeuery : "",
                    saveBody : "/tmp/inputtest_bin"
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "200 OK"
                }
            },{
                request : {
                    method : "GET",
                    path : "/stream.bin",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    readBody : "/tmp/inputtest_bin"
                }
            
            },{
                request : {
                    method : "PUT",
                    path : "/stream.txt",
                    qeuery : "",
                    saveBody : "/tmp/inputtest"
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "200 OK"
                }
            },{
                request : {
                    method : "GET",
                    path : "/stream.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    readBody : "/tmp/inputtest"
                }
            
            },{
                request : {
                    method : "PUT",
                    path : "/maxaged.txt",
                    qeuery : "",
                    saveBody : "/tmp/maxaged"
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "200 OK"
                }
            },{
                request : {
                    method : "HEAD",
                    path : "/maxaged.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        'cache-control' : 'max-age=3600'
                    },
                    readBody : "/tmp/maxaged"
                }
            
            },{
                request : {
                    method : "GET",
                    path : "/source.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 302,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "location" : "/dest.txt"
                    },
                    body : "302 Moved"
                }
            },{
                request : {
                    method : "GET",
                    path : "/redirection_bad.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 302,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "302 Moved"
                }
            },{
                request : {
                    method : "GET",
                    path : "/dest.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "Set-Cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "X-Yhoo-Debug" : "true"
                    },
                    body : "200 Ok"
                }
            
            },{
                request : {
                    method : "POST",
                    path : "/mbst/admin/v1/tokenized_url",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        'x-ysws-token-expiry-date' : 'Tue, 9 Aug 2011 20:29:34 GMT',
                        'content-type' : 'text/plain',
                        'cache-control' : 'no-cache',
                        date : 'Tue, 09 Aug 2011 17:42:54 GMT',
                        age : '0',
                        via : 'HTTP/1.1 web2.tusw1.mobstor.sp2.yahoo.com (YahooTrafficServer/1.19.8 [c sSf ])',
                        server : 'YTS/1.19.4',
                        'x-ysws-request-id' : 'f503a720-1a14-4cec-b741-30c4b12cd099',
                        connection : 'close'
                    },
                    body : "playground.yahoofs.com/mbst/tokenized_url/v1/867157656/VEVUV2djM2pmRU9BOXdDVDU4cUdoUkkwTFp6dVF4U0V5cFVTNnlCeVMxUTFnU1FPdHVMWUZ1elc0QmtCcnpVRjd1bHE0S0hLUE1MOUtNRXczbVRhRXprcUVwU2Q3eVdTcXN1UXpFMmV1ZElWNGRWYWUyQmd6Q2RCWUNNVzBieDdaeVUt"
                }
            },{
                request : {
                    method : "PUT",
                    path : "/test.txt",
                    qeuery : "",
                    saveBody : "/tmp/test.txt"
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "x-yhoo-debug" : "true"
                    },
                    body : "200 OK"
                }
            },{
                request : {
                    method : "PUT",
                    path : "/test_object.txt",
                    qeuery : "",
                    saveBody : "/tmp/test_object.txt"
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "x-yhoo-debug" : "true"
                    },
                    readBody : "/tmp/test_object.txt"
                }
            },{
                request : {
                    method : "GET",
                    path : "/test.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "x-yahoo-debug" : "true"
                    },
                    readBody : "/tmp/test.txt"
                }
            },{
                request : {
                    method : "HEAD",
                    path : "/test.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "x-yahoo-debug" : "true"
                    },
                    readBody : "/tmp/test.txt"
                }
            },{
                request : {
                    method : "HEAD",
                    path : "/test/",
                    qeuery : ""
                },
                response : {
                    statusCode : 200,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com",
                        "x-yahoo-debug" : "true"
                    },
                    body : "200 Ok"
                }
            },{
                request : {
                    method : "GET",
                    path : "/404_not_found.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 404,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "404 Not found"
                }
            },{
                request : {
                    method : "GET",
                    path : "/builds/build-3.1.0/videoplayer-min.js",
                    qeuery : ""
                },
                response : {
                    statusCode : 302,
                    headers : {
                        "location" : "http://gops.use12.mobstor.vip.bf1.yahoo.com/yvideo/builds/build-3.1.0/videoplayer-min.js?noredirect=1&visited=gops.usw12.mobstor.vip.gq1.yahoo.com"
                    },
                    body : "404 Not found"
                }
            },{
                request : {
                    method : "GET",
                    path : "/nocertificate.txt",
                    qeuery : ""
                },
                response : {
                    statusCode : 401,
                    headers : {
                        "set-cookie" : "CH=deleted; expires=Wed, 16-Jun-2010 18:40:30 GMT; path=/; domain=www.yahoo.com"
                    },
                    body : "401 Unauthorized"
                }
            }
        ]
    },
    'gops.use12.mobstor.vip.bf1.yahoo.com' : {
        httpVersion : '1.0',
        defaultEncoding : 'utf-8',
        storage : '/tmp',
        senario : [
            {
                request : {
                    method : "GET",
                    path : "/yvideo/builds/build-3.1.0/videoplayer-min.js",
                    qeuery : ""
                },
                response : {
                    statusCode : 404,
                    headers : { },
                    body : "404 Not found"
                }
            }
        ]
    }
};
